lib/actions/usbsdfuTargetActions.js

```diff
@@ -268,103 +287,196 @@ function calculateSHA512Hash(image) {
+                .setIn(['initPacket', 'hash'], hash);
+}
 
-        dispatch(targetActions.writeProgressStartAction());
-        dfu.start(true)
+// Operate DFU process with update all the images on the target device
+function operateDFU(inputDfuImages) {
+    if (inputDfuImages.size === 0) {
+        logger.info('All dfu images have been written to the target device.');
+        return dispatch => Promise.resolve()
             .then(() => {
-                logger.info('DFU completed successfully!');
                 dispatch(targetActions.writeProgressFinishedAction());
-                dispatch({ type: 'DEVICE_DESELECTED' });
-            })
-            .catch(error => {
-                logger.error(`DFU failed. Reason: ${error.message || error}`);
-            })
-            .then(() => {
-                port.close();
-                logger.info(`Port is closed: ${port.path}`);
+                dispatch(canWrite());
             });
+    }
+
+    const dfuImages = inputDfuImages;
+    let image;
+    let serialNumber;
+
+    return async (dispatch, getState) => Promise.resolve()
+        // When dfu starts
+        .then(async () => {
+            serialNumber = getState().app.target.serialNumber;
+            image = dfuImages.first();
+            const packet = await initPacket.createInitPacketUint8Array(image.initPacket);
+            const firmwareUpdates = new dfujs.DfuUpdates([{
+                initPacket: packet,
+                firmwareImage: image.firmwareImage,
+            }]);
+            const serialUsbTransport = new dfujs.DfuTransportUsbSerial(serialNumber);
+            const dfu = new dfujs.DfuOperation(firmwareUpdates, serialUsbTransport);
+
+            dispatch(targetActions.targetWritableKnownAction(false));
+            dispatch(targetActions.writeProgressStartAction());
+            logger.info('Performing DFU. This may take a few seconds');
+
+            return dfu.start(true);
+        })
+        // Catch error during dfu
+        .catch(error => {
+            logger.error(`DFU failed. Reason: ${error.message || error}`);
+            throw error;
+        })
+        // Wait for the device to reattach
+        .then(() => {
+            logger.info(`DFU for ${dfuImages.first().name} completed successfully!`);
+            logger.info(`${dfuImages.size - 1} dfu pakcage(s) left.`);
+            logger.info('Waiting for device');
+            if (image.name.includes(RegionName.BOOTLOADER) ||
+                image.name.includes(RegionName.SOFTDEVICE)) {
+                return waitForDevice(serialNumber, 5000, ['serialport']);
+            } else if (image.name.includes(RegionName.APPLICATION)) {
+                return waitForDevice(serialNumber, 5000, ['nordicusb']);
+            }
+            throw new Error('Device type is unknown when waiting for device.');
+        })
+        // Select and wait for device to open
+        .then(device => dispatch(selectAndWaitForDeviceOpen(device)))
```

Comment:
This doesn't feel right. This action only waits 500ms, which might not be enough and the consequences of triggering a `DEVICE_SELECTED` may be out of control. I suggest to expose core API functionality to disable the _DeviceLister_ under _DeviceSelector_ temporarily. That requires core change, but that's the good solution.

Reply:
I have tried 100ms, 500ms, 1000ms, 5000ms on Windows, macOS and linux virtual machine. 500ms works fine so I choose 500ms here. But it will fail on a slow pc for sure. We can discuss the change in core.

Reply:
Let's investigate how to properly expose `startWatchingDevices()` and `stopWatchingDevices()` in core API.
https://github.com/NordicSemiconductor/pc-nrfconnect-core/blob/78ecae95593d95eac4e906aba5675b9e0a4d7ac7/lib/windows/app/actions/deviceActions.js#L213
This needs to be a PR along with wiki update about the exposed functionality.
When that is completed programmer needs to block core's lister for the duration of all DFUs.